activesupport/lib/active_support/message_encryptor.rb

```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
     end
 
     # Encrypt and sign a message. We need to sign the message in order to avoid
     # padding attacks. Reference: http://www.limited-entropy.com/padding-oracle-attacks.
-    def encrypt_and_sign(value)
-      verifier.generate(_encrypt(value))
+    def encrypt_and_sign(value, options = {})
+      @claims = Claims.new(payload: value, **options)
+      _encrypt(@claims.to_h)
     end
 
     # Decrypt and verify a message. We need to verify the message in order to
     # avoid padding attacks. Reference: http://www.limited-entropy.com/padding-oracle-attacks.
-    def decrypt_and_verify(value)
-      _decrypt(verifier.verify(value))
+    def decrypt_and_verify(message, options = {})
+      if message.include? '--'
+        legacy_encryptor = LegacyMessageEncryptor.new(@secret, @sign_secret, @options)
+        legacy_encryptor.decrypt_and_verify(message)
+      else
+        if message.count('.') == 4
+          claims = _decrypt(message).symbolize_keys
+          Claims.verify!(claims, options)
+        else
+          raise InvalidFormat, 'Invalid JWT Format.'
+        end
+      end
     end
 
     private
+      def _encrypt(value)
+        protected_header = @serializer.dump header
+        cipher.encrypt
+        cipher.key = content_encryption_key
+        iv = cipher.random_iv
+        cipher.iv = iv
+        cipher.auth_data = to_ascii encode(protected_header)
+
+        ciphertext = cipher.update(@serializer.dump value) + cipher.final
+        auth_tag = cipher.auth_tag
+
+        ([protected_header, jwe_encrypted_key, iv, ciphertext, auth_tag].map { |a| encode a }).join('.')
+      end
 
-    def _encrypt(value)
-      cipher = new_cipher
-      cipher.encrypt
-      cipher.key = @secret
+      def _decrypt(encrypted_message)
+        protected_header, _encrypted_key, iv, ciphertext, auth_tag = encrypted_message.split('.').map { |a| decode a }
+        valid_header?(@serializer.load protected_header)
 
-      # Rely on OpenSSL for the initialization vector
-      iv = cipher.random_iv
+        cipher.decrypt
+        cipher.key = decrypt_content_encryption_key
+        cipher.iv = iv
+        cipher.auth_tag = auth_tag
+        cipher.auth_data = to_ascii encode(protected_header)
 
-      encrypted_data = cipher.update(@serializer.dump(value))
-      encrypted_data << cipher.final
+        decrypted_message = cipher.update(ciphertext) + cipher.final
 
-      "#{::Base64.strict_encode64 encrypted_data}--#{::Base64.strict_encode64 iv}"
-    end
+        @serializer.load decrypted_message
 
-    def _decrypt(encrypted_message)
-      cipher = new_cipher
-      encrypted_data, iv = encrypted_message.split("--").map {|v| ::Base64.strict_decode64(v)}
+      rescue OpenSSLCipherError, TypeError, ArgumentError
+        raise InvalidMessage
+      end
 
-      cipher.decrypt
-      cipher.key = @secret
-      cipher.iv  = iv
+      def cipher
+        @cipher ||= OpenSSL::Cipher.new(cipher_name)
```

Comment:
I don't know if it's safe to cache the cipher. Perhaps we should look more to what I did in Tokie with the custom cipher class: https://github.com/kaspth/tokie/blob/master/lib/tokie/encryptor/cipher.rb

That keeps all the nitty gritty encryption stuff in its own place.